# AIDM7380 Group Project - Recommender System for eLearning

# Data Exploration and Analysis

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import plotly.graph_objs as go

## Load dataset

In [2]:
elearning = pd.read_csv("elearning_dataset1.csv")
elearning

Date  UserID  CourseID                 Event
0      2022-01-10 10:12:35     440       247           view_course
1      2022-01-10 10:22:28    1186       338           view_course
2      2022-01-10 10:23:13    1186       338  detailed_description
3      2022-01-10 10:42:18     703       238           view_course
4      2022-01-10 10:51:02     225       350           view_course
...                    ...     ...       ...                   ...
17360  2022-03-09 15:09:38    1081       245           view_course
17361  2022-03-09 15:11:02     374       146  detailed_description
17362  2022-03-09 15:12:03    1081       245       teacher_profile
17363  2022-03-09 15:20:15    1142       231           view_course
17364  2022-03-09 15:23:13    1142       231       teacher_profile

[17365 rows x 4 columns]

## Check the dataset

In [3]:
users = elearning.UserID.unique()
content = elearning.CourseID.unique()
events = elearning.Event.unique()
print(f"There are {len(users)} unique users, {len(content)} unique content and {len(events)} events in the dataset.")

There are 1591 unique users, 396 unique content and 5 events in the dataset.


In [4]:
elearning.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17365 entries, 0 to 17364
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Date      17365 non-null  object
 1   UserID    17365 non-null  int64 
 2   CourseID  17365 non-null  int64 
 3   Event     17365 non-null  object
dtypes: int64(2), object(2)
memory usage: 542.8+ KB


In [5]:
elearning.Event.unique()

array(['view_course', 'detailed_description', 'teacher_profile',
       'rundown', 'institution'], dtype=object)

In [6]:
elearning.Event.describe()

count           17365
unique              5
top       view_course
freq             8874
Name: Event, dtype: object

There are five different user behaviors in this dataset, "view_course", "detailed_description", "teacher_profile", "teacher_profile", "rundown", and "institution", with the most "view_course" 8874 times.

In [7]:
# Calculate the count of each behavior using value_counts
behavior_counts = elearning['Event'].value_counts()

# Create pie chart data
data = go.Pie(labels=behavior_counts.index, values=behavior_counts.values)

# Create layout
layout = go.Layout(title='Event Counts by Behavior')

# Create the chart
fig = go.Figure(data=data, layout=layout)

# Show the chart
fig.show()

"view_course" accounted for 51.1% of all user behaviors, followed by "rundown" at 15.4%.

In [8]:
# Calculate the count of events for each CourseID using value_counts
course_event_counts = elearning['CourseID'].value_counts()

# Find the CourseID with the most events and its corresponding event count
most_event_course_id = course_event_counts.idxmax()
most_event_count = course_event_counts.max()

# Display the results
print(f"CourseID with the most events: {most_event_course_id}")
print(f"Event count: {most_event_count}")

CourseID with the most events: 202
Event count: 101


In [9]:
# Calculate the count of each CourseID using value_counts
course_counts = elearning['CourseID'].value_counts()

# Create table data
data = go.Table(
    header=dict(values=['CourseID', 'Event Counts']),
    cells=dict(values=[course_counts.index, course_counts.values])
)

# Create layout
layout = go.Layout(title='CourseID Counts')

# Create the chart
fig = go.Figure(data=data, layout=layout)

# Show the chart
fig.show()

The top 10 courses have user behaviors that all exceed 80 times, with the course id with the highest number of behaviors being 202.

In [10]:
# Filter the dataset for events with 'view_course' and group by CourseID to count the views
course_views = elearning[elearning['Event'] == 'view_course'].groupby('CourseID').size()

# Sort the views in descending order and select the top 10 courses
top_10_courses = course_views.sort_values(ascending=False).head(10)

# Create a DataFrame with the Course ID and the number of views
df = pd.DataFrame({'Course ID': top_10_courses.index, 'Number of Views': top_10_courses.values})

# Create an interactive table
fig = go.Figure(data=[go.Table(header=dict(values=list(df.columns)),
                               cells=dict(values=[df['Course ID'], df['Number of Views']]))])

# Set the table title
fig.update_layout(title="Top 10 Courses with Highest Views")

# Show the table
fig.show()

In [11]:
# Group the dataset by UserID and count the number of unique courses viewed by each user
user_course_count = elearning.groupby('UserID')['CourseID'].nunique()

# Sort the users by the number of unique courses viewed in descending order and select the top 10 users
top_10_users = user_course_count.sort_values(ascending=False).head(10)

# Create a DataFrame with the User ID and the number of unique courses viewed
df = pd.DataFrame({'User ID': top_10_users.index, 'Number of Unique Courses': top_10_users.values})

# Create an interactive table
fig = go.Figure(data=[go.Table(header=dict(values=list(df.columns)),
                               cells=dict(values=[df['User ID'], df['Number of Unique Courses']]))])

# Set the table title
fig.update_layout(title="Top 10 Users with Most Unique Course Views")

# Show the table
fig.show()

Users with user IDs 429, 1040, and 347 all viewed 14 different courses.

In [12]:
# Calculate the count of each UserID using value_counts
userid_counts = elearning['UserID'].value_counts()

# Create table data
data = go.Table(
    header=dict(values=['UserID', 'Count']),
    cells=dict(values=[userid_counts.index, userid_counts.values])
)

# Create layout
layout = go.Layout(title='UserID Counts')

# Create the chart
fig = go.Figure(data=data, layout=layout)

# Show the chart
fig.show()

User 1040 has 35 times user behaviors in this dataset, the most of any user.

In [13]:
# Calculate the count of each UserID using value_counts
userid_counts = elearning['UserID'].value_counts()

# Create histogram data
data = go.Histogram(x=userid_counts.values)

# Create layout
layout = go.Layout(title='User Behavior Counts Distribution', xaxis=dict(title='Counts'), yaxis=dict(title='Frequency'))

# Create the chart
fig = go.Figure(data=data, layout=layout)

# Show the chart
fig.show()

As can be seen from the distribution of counts in the above graph, most users have a number of behaviors between 5 and 20.

# User-based Collaborative Filtering

## Implicit Ratings

In [14]:
# Create a user-item binary matrix
uiMatrix = pd.DataFrame(columns=content, index=users)
uiMatrix.head()

247  338  238  350  264  139  9    20   8    288  ...  341  329  73   \
440   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
1186  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
703   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
225   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
1334  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   

      115  100  89   50   306  199  318  
440   NaN  NaN  NaN  NaN  NaN  NaN  NaN  
1186  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
703   NaN  NaN  NaN  NaN  NaN  NaN  NaN  
225   NaN  NaN  NaN  NaN  NaN  NaN  NaN  
1334  NaN  NaN  NaN  NaN  NaN  NaN  NaN  

[5 rows x 396 columns]

## Type of events and weights

In [15]:
eventTypes = elearning.Event.unique()
print(eventTypes)

['view_course' 'detailed_description' 'teacher_profile' 'rundown'
 'institution']


In [16]:
eventWeights = {
    'view_course': 5,
    'detailed_description': 15,
    'teacher_profile': 50,
    'institution': 50,
    'rundown': 80}

## Add Decay
Using the formula introduced during lecture

$${IRDecay}_{(i,u)} = \sum_{i=1}^n w_i*{\#event}_i*d\left({\#event}_i\right) = \left(w_1*{\#event}_1*d\left({\#event}_1\right)\right)+\left(w_2*{\#event}_2*d\left({\#event}_2\right)\right)+\dots+\left(w_n*{\#event}_n*d\left({\#event}_n\right)\right)$$

In [17]:
#Computing decay
import datetime
from datetime import date, timedelta, datetime

def compute_decay(eventDate, decayDays):
    age = (date.today() - datetime.strptime(eventDate, '%Y-%m-%d %H:%M:%S').date()) // timedelta(days=decayDays)
    #Converts the given event date string into a datetime object.Extract the date part of a datetime object.
    #print("Age of event:", age)
    decay = 1/age #simple decay
    #print("Decay factor:", decay)

    return decay

createdEvent = elearning.at[0,'Date']
thresholdDays = 2 # Number of days
decayFactor = compute_decay(createdEvent, thresholdDays)

print(decayFactor)

0.0024271844660194173


In [18]:
#Compute the Implicit Rating for each user-item combination. Populate the user-item matrix  
#Add the decay factor
for index, row in elearning.iterrows():
    currentUser = row['UserID']
    currentContent = row['CourseID']

    w = eventWeights[row['Event']] * compute_decay(row['Date'],thresholdDays)

    currentValue = uiMatrix.at[currentUser, currentContent]
    if np.isnan(currentValue):
        currentValue = 0

    updatedValue = currentValue + w #+ (1 * w)
    uiMatrix.at[currentUser, currentContent] = updatedValue

uiMatrix.head()

247       338       238       350       264  139  9    20   8    \
440   0.012136       NaN       NaN       NaN       NaN  NaN  NaN  NaN  NaN   
1186       NaN  0.048544       NaN       NaN       NaN  NaN  NaN  NaN  NaN   
703        NaN       NaN  0.012136       NaN       NaN  NaN  NaN  NaN  NaN   
225        NaN       NaN       NaN  0.012136       NaN  NaN  NaN  NaN  NaN   
1334       NaN       NaN       NaN       NaN  0.012136  NaN  NaN  NaN  NaN   

      288  ...  341  329  73        115       100  89   50   306  199  318  
440   NaN  ...  NaN  NaN  NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  
1186  NaN  ...  NaN  NaN  NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  
703   NaN  ...  NaN  NaN  NaN       NaN  0.137157  NaN  NaN  NaN  NaN  NaN  
225   NaN  ...  NaN  NaN  NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  
1334  NaN  ...  NaN  NaN  NaN  0.140665       NaN  NaN  NaN  NaN  NaN  NaN  

[5 rows x 396 columns]

## Normalization

In [19]:
#Update the user-item matrix by normalizing the values between 0 and 10.
uiMatrixNorm = uiMatrix.apply(
    lambda x: ((x - np.nanmin(uiMatrix.values))/(np.nanmax(uiMatrix.values) - np.nanmin(uiMatrix.values)))*10
    )

uiMatrixNorm.head()

247      338  238  350  264  139  9    20   8    288  ...  341  329  \
440   0.0      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
1186  NaN  0.55484  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
703   NaN      NaN  0.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
225   NaN      NaN  NaN  0.0  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
1334  NaN      NaN  NaN  NaN  0.0  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   

      73        115       100  89   50   306  199  318  
440   NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  
1186  NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  
703   NaN       NaN  1.905273  NaN  NaN  NaN  NaN  NaN  
225   NaN       NaN       NaN  NaN  NaN  NaN  NaN  NaN  
1334  NaN  1.958732       NaN  NaN  NaN  NaN  NaN  NaN  

[5 rows x 396 columns]

## User-based Function
Step1: Compute Similarity between the active user and the rest of the users

Step2: Select the items whose values are to be predicted for current user

Step3: Compute the predicted ratings for current user.
Using mean value of the other high-similarity users

In [98]:
from scipy.spatial.distance import jaccard, cosine
from scipy.stats import spearmanr

def userCF_prediction(df, currentUser, numUsers, numItems, similarity='pearson',displayed = True):#display means printing outcomes or not
    df = df.astype(float)
    df_copy = df.copy()
    cuDf = df_copy.loc[currentUser]

    if similarity == 'pearson':
        similarityDf = df_copy.corrwith(cuDf, axis=1, method='pearson')
    elif similarity == 'jaccard':
        similarityDf = df_copy.apply(lambda row: 1 - jaccard(cuDf, row), axis=1)
    elif similarity == 'cosine':
        similarityDf = df_copy.apply(lambda row: 1 - cosine(cuDf, row), axis=1)
    elif similarity == 'spearman':
        similarityDf = df_copy.apply(lambda row: spearmanr(cuDf, row).correlation, axis=1)
    else:
        raise ValueError("Invalid similarity measure. Available options are 'pearson', 'jaccard', 'cosine', and 'spearman'.")

    similarityDf.sort_values(ascending=False, inplace=True)
    similarityDf.drop(labels=[currentUser], inplace=True)
    similarityDf = similarityDf.head(numUsers)

    toPredict = cuDf[cuDf.isna()]

    ratings = df_copy.loc[similarityDf.index]
    ratingsToPredict = ratings[toPredict.index]

    predictedRatings = ratingsToPredict.mean()
    predictedRatings.sort_values(ascending=False, inplace=True)
    if displayed:
        display(predictedRatings.head(numItems))
    return predictedRatings.head(numItems)

In [99]:
#Find one without all the ratings already filled to recommend something for him
uiMatrixNorm.isna().sum(axis=1)

440     391
1186    390
703     392
225     390
1334    391
       ... 
659     395
66      395
612     395
483     395
956     395
Length: 1591, dtype: int64

### Pearson Showcase

In [100]:
recommendList1 = userCF_prediction(uiMatrixNorm,440,200,5,'pearson')

d:\Tools\anaconda2024\Lib\site-packages\numpy\lib\function_base.py:2889: RuntimeWarning:

Degrees of freedom <= 0 for slice

d:\Tools\anaconda2024\Lib\site-packages\numpy\lib\function_base.py:2748: RuntimeWarning:

divide by zero encountered in divide

d:\Tools\anaconda2024\Lib\site-packages\numpy\lib\function_base.py:2748: RuntimeWarning:

invalid value encountered in multiply



326    6.725168
259    5.616933
196    5.010373
236    5.010373
186    4.945594
dtype: float64

In [101]:
recommendList2 = userCF_prediction(uiMatrixNorm,823,100,5,'pearson')

d:\Tools\anaconda2024\Lib\site-packages\numpy\lib\function_base.py:2889: RuntimeWarning:

Degrees of freedom <= 0 for slice

d:\Tools\anaconda2024\Lib\site-packages\numpy\lib\function_base.py:2748: RuntimeWarning:

divide by zero encountered in divide

d:\Tools\anaconda2024\Lib\site-packages\numpy\lib\function_base.py:2748: RuntimeWarning:

invalid value encountered in multiply



359    5.616933
79     5.431615
42     5.063387
236    5.010373
33     4.984266
dtype: float64

### Spearman Showcase

In [102]:
recommendList3 = userCF_prediction(uiMatrixNorm,440,200,5,'spearman')

326    6.725168
259    5.616933
236    5.010373
196    5.010373
186    4.945594
dtype: float64

In [103]:
recommendList4 = userCF_prediction(uiMatrixNorm,823,100,5,'spearman')

359    5.616933
79     5.431615
42     5.063387
236    5.010373
214    4.984266
dtype: float64

In [117]:
#Delete NaN values and return new list
def process_recommendations(recommendations,displayed=True):
    filtered_series = recommendations.dropna()
    if(displayed):
        print('CurrentLength:'+str(len(filtered_series)))
        display(filtered_series)
    return filtered_series

## Example

In [118]:
recommendListExample = userCF_prediction(uiMatrixNorm,2,2,50,'spearman')
filtered_series = process_recommendations(recommendListExample)

248    4.971311
323    3.532030
381    2.989971
353    0.590606
181    0.580863
114    0.577034
106    0.556640
386    0.556640
338    0.554840
384    0.000902
247    0.000000
238         NaN
350         NaN
264         NaN
139         NaN
9           NaN
20          NaN
8           NaN
288         NaN
373         NaN
298         NaN
180         NaN
224         NaN
150         NaN
30          NaN
3           NaN
284         NaN
235         NaN
261         NaN
117         NaN
177         NaN
48          NaN
90          NaN
291         NaN
119         NaN
362         NaN
252         NaN
136         NaN
255         NaN
321         NaN
258         NaN
138         NaN
374         NaN
360         NaN
98          NaN
86          NaN
2           NaN
349         NaN
201         NaN
233         NaN
dtype: float64

CurrentLength:11


248    4.971311
323    3.532030
381    2.989971
353    0.590606
181    0.580863
114    0.577034
106    0.556640
386    0.556640
338    0.554840
384    0.000902
247    0.000000
dtype: float64

# Item-based Collaborative Filtering

Use the cosine distance in scipy library to compute the similarity.

$$cosine\_similarity = 1 - cosine\_distance$$

In [26]:
from scipy.spatial.distance import cosine

def cosine_sim(df1, df2):
    # check for na in dataframes
    df1na = df1.isna()
    df1clean = df1[~df1na]
    df2clean = df2[~df1na]

    df2na = df2clean.isna()
    df1clean = df1clean[~df2na]
    df2clean = df2clean[~df2na]


    # Compute cosine similarity
    distance = cosine(df1clean, df2clean)
    sim = 1 - distance

    return sim

## Precomputing similarities

A utility function to convert from numeric to boolean, indicating if the user rated an item.

In [27]:
def to_bool(value):
    if np.isnan(value):
        return 0
    else:
        return 1

Compute the number of overlapping rating between each item

In [28]:
uiMatrixBool = uiMatrix.applymap(lambda x: to_bool(x))
overlappingUsersRatings = uiMatrixBool.T.dot(uiMatrixBool)

C:\Users\罗萍\AppData\Local\Temp\ipykernel_21284\243225764.py:1: FutureWarning:

DataFrame.applymap has been deprecated. Use DataFrame.map instead.



In [29]:
display(overlappingUsersRatings)

247  338  238  350  264  139  9    20   8    288  ...  341  329  73   \
247   45    1    0    0    2    0    2    0    0    0  ...    0    0    0   
338    1   33    0    1    2    2    3    0    1    1  ...    0    0    0   
238    0    0   21    0    0    1    0    1    0    0  ...    0    0    0   
350    0    1    0   28    0    1    0    0    1    1  ...    0    0    0   
264    2    2    0    0   29    0    0    0    0    1  ...    0    0    0   
..   ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
89     2    0    0    0    1    0    0    0    0    0  ...    0    0    0   
50     0    0    0    0    0    0    0    0    0    0  ...    0    0    0   
306    1    0    0    0    0    0    0    0    0    0  ...    0    0    0   
199    0    0    0    0    0    0    0    0    1    0  ...    0    0    0   
318    1    0    0    0    0    0    0    0    0    0  ...    0    0    0   

     115  100  89   50   306  199  318  
247    1    2    2    0    1    0    1  
338    0    0    0    0    0    0    0  
238    0    1    0    0    0    0    0  
350    0    0    0    0    0    0    0  
264    1    1    1    0    0    0    0  
..   ...  ...  ...  ...  ...  ...  ...  
89     0    0    7    0    0    0    0  
50     0    0    0    1    0    0    0  
306    0    0    0    0    2    0    0  
199    0    0    0    0    0    2    0  
318    0    1    0    0    0    0    1  

[396 rows x 396 columns]

Check min and max

In [30]:
overlappingUsersRatings.max(axis=0).max()

48

In [31]:
overlappingUsersRatings.min(axis=0).min()

0

In [32]:
toDrop = overlappingUsersRatings.max(axis=0) > 10
display(toDrop)

247     True
338     True
238     True
350     True
264     True
       ...  
89     False
50     False
306    False
199    False
318    False
Length: 396, dtype: bool

In [33]:
selectedItems = overlappingUsersRatings.loc[toDrop, toDrop]
display(selectedItems)

247  338  238  350  264  139  9    20   288  373  ...  300  27   167  \
247   45    1    0    0    2    0    2    0    0    0  ...    1    0    1   
338    1   33    0    1    2    2    3    0    1    0  ...    2    0    0   
238    0    0   21    0    0    1    0    1    0    0  ...    1    0    0   
350    0    1    0   28    0    1    0    0    1    0  ...    0    0    0   
264    2    2    0    0   29    0    0    0    1    0  ...    0    0    1   
..   ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
28     1    0    0    0    0    0    0    0    0    0  ...    0    0    0   
23     0    0    0    0    1    0    1    0    0    0  ...    0    0    0   
121    1    0    0    1    1    1    0    1    1    0  ...    0    0    0   
128    1    0    0    0    0    1    1    1    0    1  ...    0    0    1   
100    2    0    1    0    1    1    0    0    0    0  ...    0    0    0   

     290  76   28   23   121  128  100  
247    0    0    1    0    1    1    2  
338    1    0    0    0    0    0    0  
238    0    0    0    0    0    0    1  
350    0    0    0    0    1    0    0  
264    0    1    0    1    1    0    1  
..   ...  ...  ...  ...  ...  ...  ...  
28     0    0   14    1    0    0    0  
23     0    0    1   19    0    0    0  
121    0    0    0    0   22    0    0  
128    0    0    0    0    0   12    1  
100    0    1    0    0    0    1   14  

[353 rows x 353 columns]

Compute item-item similarity

In [34]:
selectedIndex = selectedItems.index
uiMatrixSelection = uiMatrix[selectedIndex]
uiMatrixSelection = uiMatrixSelection.sub(uiMatrixSelection.mean(axis=1), axis=0)
iiSimMatrix = pd.DataFrame().reindex_like(selectedItems)
for item1 in selectedIndex:
    item1Ratings = uiMatrixSelection[item1]
    for item2 in selectedIndex:
        item2Ratings = uiMatrixSelection[item2]
        if overlappingUsersRatings.at[item1, item2] > 1:
          iiSimMatrix.at[item1, item2] = cosine_sim(item1Ratings, item2Ratings)


In [35]:
display(iiSimMatrix)

247       338  238  350       264       139       9    20   288  \
247  1.000000       NaN  NaN  NaN  0.904515       NaN  0.557043  NaN  NaN   
338       NaN  1.000000  NaN  NaN  0.346367 -0.860503 -0.587661  NaN  NaN   
238       NaN       NaN  1.0  NaN       NaN       NaN       NaN  NaN  NaN   
350       NaN       NaN  NaN  1.0       NaN       NaN       NaN  NaN  NaN   
264  0.904515  0.346367  NaN  NaN  1.000000       NaN       NaN  NaN  NaN   
..        ...       ...  ...  ...       ...       ...       ...  ...  ...   
28        NaN       NaN  NaN  NaN       NaN       NaN       NaN  NaN  NaN   
23        NaN       NaN  NaN  NaN       NaN       NaN       NaN  NaN  NaN   
121       NaN       NaN  NaN  NaN       NaN       NaN       NaN  NaN  NaN   
128       NaN       NaN  NaN  NaN       NaN       NaN       NaN  NaN  NaN   
100 -0.871078       NaN  NaN  NaN       NaN       NaN       NaN  NaN  NaN   

     373  ...      300  27   167  290  76   28   23   121  128       100  
247  NaN  ...      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN -0.871078  
338  NaN  ... -0.10207  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN       NaN  
238  NaN  ...      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN       NaN  
350  NaN  ...      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN       NaN  
264  NaN  ...      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN       NaN  
..   ...  ...      ...  ...  ...  ...  ...  ...  ...  ...  ...       ...  
28   NaN  ...      NaN  NaN  NaN  NaN  NaN  1.0  NaN  NaN  NaN       NaN  
23   NaN  ...      NaN  NaN  NaN  NaN  NaN  NaN  1.0  NaN  NaN       NaN  
121  NaN  ...      NaN  NaN  NaN  NaN  NaN  NaN  NaN  1.0  NaN       NaN  
128  NaN  ...      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  1.0       NaN  
100  NaN  ...      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  1.000000  

[353 rows x 353 columns]

## Item score predict function 
based on precomputed similarity matrix

In [36]:
def item_predict(df, currentUser, currentItem, simMatrix, k):
    # df: the UI matrix dataframe ; currentUser currentItem: the current user and item for whom the prediction is to be made
    # simMatrix: the precomputed simMatrix ; k: for top-k
    
    uiMatrixNorm = df.astype(float)
    uiMatrixSelection = uiMatrixNorm.dropna(subset = [currentItem])
    cuAvgRating = uiMatrixNorm.loc[[currentUser]].dropna(axis=1).mean(axis=1)

    # Subtract the average rating of each user from that user's rating to eliminate the effect of different rating scales for different users
    uiMatrixSelection = uiMatrixSelection.sub(uiMatrixSelection.mean(axis=1), axis=0)
    uiMatrixSelection.drop(columns=[currentItem], inplace=True)
    iiSimilarity = simMatrix.loc[currentItem].dropna().sort_values(ascending=False).drop(index=currentItem)

    # Select the top k items with the highest similarity
    itemsToCompare = uiMatrixSelection[iiSimilarity.head(k).index]
    predictedRatings = itemsToCompare.mean(axis=0)
    predictedRatings.sort_values(ascending=False, inplace=True)

    # Add the average user rating back to the predicted rating
    predictedRatings += cuAvgRating.iloc[0]

    # The average of all predicted ratings is used as the final predicted rating of the current user for the current item
    return predictedRatings.mean()


## Item-based Recommand Function


In [37]:
def itemCF_recommend(df, currentUser, simMatrix, numItems, k):
    # df: the UI matrix dataframe ; currentUser: the current user for whom the recommand is to be made
    # simMatrix: the precomputed simMatrix ; numItems: the number of items to be recommanded ; k: for top-k

    # Get items not rated by the current user
    cuRatedItems = df.loc[currentUser].dropna().sort_values(ascending=False)
    unRated = df.loc[currentUser]
    unRated.drop(labels=cuRatedItems.index, inplace=True)

    toRec = []
    # Iterate through the unrated items and call the `item_predict` function for each item to predict the rating
    for item in unRated.index:
        if item in simMatrix.index:
            preRate = item_predict(df, currentUser, item, simMatrix,k)
            if not np.isnan(preRate):
                toRec.append((preRate, item))

    sortedRec = sorted(toRec, key=lambda x: x[0], reverse=True)
    return sortedRec[0:numItems]


## Tests and parameter change tests

Recommand for user 429, with numItems is 10 and k is 5 :

In [38]:
recSer = itemCF_recommend(uiMatrixNorm,429,iiSimMatrix,10,5)
for item in recSer:
    print(str(item[1])+'\t'+str(item[0]))

110	2.884301850342542
28	2.7158740000178545
2	2.692071893096693
162	2.6540392068443746
158	2.6504712620645328
147	2.6336721755609753
333	2.6136527654705954
35	2.6005071697775097
220	2.596277661817805
317	2.5952412235326348


Recommand for user 429, with numItems is 10 and k is 10 :

In [39]:
recSer = itemCF_recommend(uiMatrixNorm,429,iiSimMatrix,10,10)
for item in recSer:
    print(str(item[1])+'\t'+str(item[0]))

2	2.692071893096693
333	2.578829557642099
63	2.5328154924712574
74	2.420071825391795
45	2.3894465716108337
377	2.387426011113214
365	2.3758519217306757
60	2.3505283419395107
309	2.3392888090641417
382	2.3326332886866354


Recommand for user 347 with numItems is 10 and k is 5 :

In [40]:
recSer = itemCF_recommend(uiMatrixNorm,347,iiSimMatrix,10,5)
for item in recSer:
    print(str(item[1])+'\t'+str(item[0]))

110	2.427241499998883
32	2.3906488537429733
283	2.30377379626116
28	2.2588136496741953
2	2.235011542753033
162	2.196978856500715
158	2.193410911720873
147	2.176611825217315
333	2.1565924151269353
35	2.14344681943385


Recommand for user 347 with numItems is 10 and k is 10 :

In [41]:
recSer = itemCF_recommend(uiMatrixNorm,347,iiSimMatrix,10,10)
for item in recSer:
    print(str(item[1])+'\t'+str(item[0]))

2	2.235011542753033
333	2.1217692072984393
63	2.075755142127597
74	1.9630114750481353
377	1.9303656607695545
32	1.9196399588257793
365	1.9187915713870156
60	1.8934679915958508
309	1.882228458720482
382	1.8755729383429756


# 给@康英参考的User Coverage函数
基本照着你微信里面的代码搭建的，其他边边角角变量名啥的你自己圆一下，调不明白再找我

In [123]:
def get_recommended_users(df,numUsers,numItems,similarity='pearson', displayed=False):

    users = df.index.unique()

    recommended_users = []

    for user_id in users:
        recommendations = userCF_prediction(df,user_id,numUsers,numItems,similarity,displayed)
        filtered_series = process_recommendations(recommendations,displayed)
        if len(filtered_series) >= numItems:
            recommended_users.append(user_id)

    user_coverage = len(recommended_users) / len(users)
    print("recommended_users:")
    display(recommended_users)
    print("user_coverage:")
    display(user_coverage)
    return recommended_users, user_coverage

按照我下面这个方法传入你要测试的参数就能直接出结果，参数传啥你自己看

Usecoverage = 0 的话有可能是人太少但是物品要求太多，即没有哪位用户能推到这么多电影

我这个跑一次要2-3分钟，电脑差一点要更久，5分钟还不出来可能就卡死了，重新运行一下

In [124]:
recommended_users, user_coverage=get_recommended_users(uiMatrixNorm,3,12,'spearman',False)#df,numUser,numItem,similarity function, display outcome or not

recommended_users:


[440,
 1186,
 703,
 225,
 1334,
 777,
 651,
 1467,
 1077,
 1215,
 1225,
 1401,
 758,
 988,
 1493,
 1149,
 195,
 1598,
 833,
 1555,
 1528,
 1155,
 200,
 352,
 245,
 1239,
 767,
 253,
 691,
 77,
 434,
 339,
 75,
 1449,
 897,
 566,
 222,
 974,
 311,
 1301,
 494,
 161,
 724,
 914,
 992,
 854,
 1358,
 529,
 99,
 532,
 1473,
 583,
 233,
 501,
 595,
 1368,
 706,
 565,
 1020,
 675,
 180,
 879,
 1092,
 538,
 904,
 1346,
 468,
 488,
 1549,
 587,
 1198,
 102,
 272,
 1081,
 1185,
 288,
 302,
 118,
 798,
 862,
 260,
 1369,
 867,
 620,
 460,
 528,
 328,
 888,
 1264,
 357,
 401,
 73,
 422,
 314,
 1138,
 950,
 940,
 1269,
 1546,
 577,
 382,
 836,
 145,
 209,
 1104,
 217,
 907,
 838,
 353,
 775,
 387,
 95,
 1139,
 1122,
 156,
 1267,
 820,
 1419,
 1595,
 208,
 951,
 1004,
 628,
 154,
 1307,
 234,
 406,
 263,
 649,
 105,
 685,
 948,
 844,
 544,
 348,
 86,
 0,
 482,
 523,
 1070,
 384,
 1075,
 456,
 822,
 1295,
 33,
 45,
 1509,
 1360,
 269,
 416,
 786,
 1270,
 9,
 700,
 1472,
 210,
 1262,
 811,
 878,
 1010

user_coverage:


0.9993714644877436